packages/common-scripts/src/deploy.ts

```diff
@@ -0,0 +1,572 @@
+
+  return new utils.CKBHasher().update(new Reader(outputData)).digestHex();
+}
+
+interface ScriptConfig {
+  // if hash_type is type, code_hash is ckbHash(type_script)
+  // if hash_type is data, code_hash is ckbHash(data)
+  CODE_HASH: string;
+
+  HASH_TYPE: "type" | "data";
+
+  TX_HASH: string;
+  // the deploy cell can be found at index of tx's outputs
+  INDEX: string;
+
+  // now deployWithX only supportted `code `
+  DEP_TYPE: "dep_group" | "code";
+
+  // empty
+  SHORT_ID?: number;
+}
+
+function calculateTxHash(txSkeleton: TransactionSkeletonType): string {
+  const tx = createTransactionFromSkeleton(txSkeleton);
+  const txHash = utils
+    .ckbHash(
+      core.SerializeRawTransaction(normalizers.NormalizeRawTransaction(tx))
+    )
+    .serializeJson();
+  return txHash;
+}
+
+function getScriptConfigByDataHash(
+  txSkeleton: TransactionSkeletonType,
+  outputIndex: number
+): ScriptConfig {
+  const data = txSkeleton.outputs.get(outputIndex)!.data;
+  const codeHash = utils
+    .ckbHash(new Reader(data).toArrayBuffer())
+    .serializeJson();
+  const txHash = calculateTxHash(txSkeleton);
+  const scriptConfig: ScriptConfig = {
+    CODE_HASH: codeHash,
+    HASH_TYPE: "data",
+    TX_HASH: txHash,
+    INDEX: "0x0",
+    DEP_TYPE: "code",
+  };
+  return scriptConfig;
+}
+
+function getScriptConfigByTypeHash(
+  txSkeleton: TransactionSkeletonType,
+  outputIndex: number
+): ScriptConfig {
+  const typeScript = txSkeleton.outputs.get(outputIndex)!.cell_output.type!;
+  const codeHash = utils.computeScriptHash(typeScript);
+  const txHash = calculateTxHash(txSkeleton);
+  const scriptConfig: ScriptConfig = {
+    CODE_HASH: codeHash,
+    HASH_TYPE: "type",
+    TX_HASH: txHash,
+    INDEX: "0x0",
+    DEP_TYPE: "code",
+  };
+  return scriptConfig;
+}
+
+function getScriptConfig(
+  txSkeleton: TransactionSkeletonType,
+  outputIndex: number
+): ScriptConfig {
+  const outputCell = txSkeleton.outputs.get(outputIndex);
+  if (outputCell == undefined)
+    throw new Error("Invalid txSkeleton or outputIndex");
+  const type = outputCell.cell_output.type;
+  if (type !== undefined)
+    return getScriptConfigByTypeHash(txSkeleton, outputIndex);
+  return getScriptConfigByDataHash(txSkeleton, outputIndex);
+}
+
+interface DeployOptions {
+  cellProvider: CellProvider;
+  scriptBinary: Uint8Array;
+  fromInfo: FromInfo;
+  config?: Config;
+}
+
+interface UpgradeOptions extends DeployOptions {
+  typeId: Script;
+}
+
+interface DeployResult {
+  txSkeleton: TransactionSkeletonType;
+  scriptConfig: ScriptConfig;
+}
+
+interface TypeIDDeployResult extends DeployResult {
+  typeId: Script;
+}
+
+/**
+ * Generate txSkeleton for writing binary data to CKB, usually for deploying contracts.
+ * This generator only supports `SECP256K1_BLAKE160` and `SECP256K1_BLAKE160_MULTISIG` currently.
+ *
+ * @param options
+ */
+export async function generateDeployWithDataTx(
+  options: DeployOptions
+): Promise<DeployResult> {
```

Comment:
We should first verify that `FromInfo` is secp256k1 or multisig, and throw an error if it is not

```ts
function isMultisigFromInfo(fromInfo: FromInfo): fromInfo is MultisigScript {
  if(typeof fromInfo !== 'object') return false;
  return "M" in fromInfo && "R" in fromInfo && Array.isArray(fromInfo.publicKeyHashes);
}

if (typeof fromInfo === 'string') {
  if(nameOfScript(parseAddress(fromInfo), config.SCRIPTS) !== 'SECP256K1_BLAKE160') throw new Error('...')
} else if (!isMultisigFromInfo(fromInfo)) {
  throw new Error('only SECP256K1_BLAKE160 or SECP256K1_MULTISIG is supported ')
} 

 
```